### PR TITLE
ref(planning): refactor outlier removal

### PR DESCRIPTION
**Pull request by lourenco31**
8693htzef

### CODE REVIEW

File: src/planning/include/planning/track.hpp [hunk 1]
```diff
@@ -41,6 +43,30 @@ class Track {
...
+  void logCones(bool side);
+
+  /**
+   * @brief rounds numbers to n decimal places
+   * 
+   * @param num number to be rounded
+   * @param decimal_places number of decimal places
+   * @return rounded number
+   */
+  float round_n(float num, int decimal_places);
```
Comment (marhcouto) [inline src/planning/include/planning/track.hpp:60]: This method is not related to this class whatsoever. As such, it should be in another place, maybe an utils file

File: src/planning/src/planning/path_smoothing.cpp [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+  gsl_multifit_linear_free(mw2);
+
+  RCLCPP_DEBUG(rclcpp::get_logger("rclcpp"),
+  "END fitSpline with %i points", static_cast<int>(path_eval.size()));
+
+  return path_eval;
+}
+
+std::vector<PathPoint *> PathSmoothing::pathSmoother(int precision,
+    int order, float coeffs_ratio, std::vector<PathPoint *> unord_path) {
```
Comment (marhcouto) [inline src/planning/src/planning/path_smoothing.cpp:214]: Any reason for this method to exist? It seems a bit of excessive delegation to have validate call this that calls spline smoother don't you think?

File: src/planning/src/planning/path_smoothing.cpp [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    vx = new_unord_path[min_index].first->getX() - path_point1->getX();
+    vy = new_unord_path[min_index].first->getY() - path_point1->getY();
+    path_point1 = new_unord_path[min_index].first;
+
+    path.push_back(new_unord_path[min_index].first);  // add path points to ordered sequence
+  }
+  return path;
+}
+
+void PathSmoothing::validate(std::vector<PathPoint *> a_path) {
```
Comment (marhcouto) [inline src/planning/src/planning/path_smoothing.cpp:95]: I know it was called validate but I don't think the name is appropriate anymore as it changes the path drastically, it isn't really a validation

File: src/planning/src/planning/track.cpp [hunk 4]
```diff
@@ -15,7 +15,59 @@ Track::~Track() {
   for (int i = 0; i < rSize; i++) delete rightCones[i];
 }
 
+void Track::logCones(bool side) {
+  if (side) {
+    std::cout << "LEFT CONES: ";
```
Comment (marhcouto) [inline src/planning/src/planning/track.cpp:20]: Any reason to use std::cout instead of logging?

File: src/planning/test/tests.cpp [hunk 5]
```diff
@@ -80,87 +96,597 @@ void outlierCalculations(std::string filename, std::string testname) {
...
-                                 PathPoint(4, 3.5),   PathPoint(4.25, 4.25), PathPoint(4.5, 5)};
-  EXPECT_EQ(expected, path);
+  std::vector<PathPoint> expected{
+    PathPoint(1.5,  0.0),  PathPoint(1.75,  1.0),  PathPoint(2.25,  1.4),
+    PathPoint(2.5,  2.4),  PathPoint(3.0,  2.75),  PathPoint(3.5,  3.15),
+    PathPoint(3.75,  3.9),  PathPoint(4.25,  4.25),  PathPoint(4.5,  5.0)};
+  EXPECT_EQ(path.size(),  expected.size());
+  EXPECT_EQ(expected,  path);
+}
+
```
Comment (marhcouto) [inline src/planning/test/tests.cpp:334]: Add doxygen documentation describing the scenario of the test

File: src/planning/test/tests.cpp [hunk 6]
```diff
@@ -1,14 +1,30 @@
...
 #include "planning/global_path_planner.hpp"
 #include "planning/local_path_planner.hpp"
+#include "planning/path_smoothing.hpp"
+#include "planning/track.hpp"
 #include "rclcpp/rclcpp.hpp"
 #include "utils/files.hpp"
 
 using testing::Eq;
 
+bool customComparator(const std::pair<double,  double>& a,  const std::pair<double,  double>& b) {
```
Comment (marhcouto) [inline src/planning/test/tests.cpp:15]: If it needs to be like this so be it, but it really seems like you could use a data type in this case i.e. instead of a vector of Pairs, use a vector of a class (like Position) that overloads the < operator. In any way, it is just my opinion, not necessarily better, just though it felt right.

File: src/planning/test/tests.cpp [hunk 7]
```diff
@@ -1,14 +1,30 @@
...
 
 #include "gtest/gtest.h"
 #include "planning/global_path_planner.hpp"
 #include "planning/local_path_planner.hpp"
+#include "planning/path_smoothing.hpp"
+#include "planning/track.hpp"
 #include "rclcpp/rclcpp.hpp"
 #include "utils/files.hpp"
 
 using testing::Eq;
```
Comment (marhcouto) [inline src/planning/test/tests.cpp:14]: Doxygen doc please

**Review by marhcouto — CHANGES_REQUESTED** [review]

File: src/planning/include/planning/track.hpp (hunk 1 above)
Comment (lourenco31) [reply src/planning/include/planning/track.hpp:60]: > This method is not related to this class whatsoever. As such, it should be in another place, maybe an utils file

Since it is used for testing only I'll place it in the tests.cpp file

File: src/planning/src/planning/path_smoothing.cpp (hunk 2 above)
Comment (lourenco31) [reply src/planning/src/planning/path_smoothing.cpp:214]: The reason behind having "validate" as an independent method is to call this function with predefined values for the first parameters without actually defining a default value for them since there's nothing special about any value, while also leaving the possibility to call "pathSmoother" with other inputs.

The reason behind having "pathSmoother" as an independent method is to maintain "orderPath" and "splineSmoother" as separate functions (increasing modularity and readability, but also leaving a chance to change only one of them) but to have a way to call them together since that's the only way in which they should be used.

File: src/planning/src/planning/path_smoothing.cpp (hunk 3 above)
Comment (lourenco31) [reply src/planning/src/planning/path_smoothing.cpp:95]: ✅

File: src/planning/src/planning/track.cpp (hunk 4 above)
Comment (lourenco31) [reply src/planning/src/planning/track.cpp:20]: This function was only used for visualizing the data with Desmos and it formats the cones' coordinates accordingly. To clarify, I left it as a method of the Track class just because it may be useful for debugging, even though it is currently only used for testing.

Comment (lourenco31) [reply src/planning/src/planning/track.cpp:20]: Logging doesn't work because either it prints each cone in one line or it truncates most of the cones because there is a maximum number of characters in each log.

File: src/planning/test/tests.cpp (hunk 6 above)
Comment (lourenco31) [reply src/planning/test/tests.cpp:15]: It definitely did not need to be like that but the formatting of the result of the Delaunay Triangulations into a vector of pairs was already implemented and it makes testing substantially easier.

File: src/planning/test/tests.cpp (hunk 5 above)
Comment (lourenco31) [reply src/planning/test/tests.cpp:385]: ✅

File: src/planning/test/tests.cpp (hunk 7 above)
Comment (lourenco31) [reply src/planning/test/tests.cpp:14]: ✅

File: src/planning/test/tests.cpp (hunk 6 above)
Comment (andrelds11) [reply src/planning/test/tests.cpp:24]: Agree on this. Why would it be easier?

Comment (marhcouto) [reply src/planning/test/tests.cpp:24]: Na, thats valid

**Review by marhcouto — APPROVED** [review]

**Review by andrelds11 — APPROVED** [review]

**Review by PedromLBarros — APPROVED** [review]
